### DISCUSSION TITLE
Investigate increasing developer productivity with pull requests environments

### DISCUSSION THREAD

**Discussion by waveywaves** [directus/directus#16026] [Archive]
I would like to make life easier for Directus maintainers by implementing [Uffizzi](https://github.com/UffizziCloud/uffizzi) previews. 
Disclaimer: I work on Uffizzi
Uffizzi is a Open Source full stack previews engine and is free for Directus. This will provide maintainers with previews of PRs made to the Directus github repo which would be deployed in the cloud. This will allow them to review and get PRs merged faster.

TODO:

- [ ] Intial POC

**Comment by rijkvanzanten** [comment]
Definitely an interesting concept, but I'm not sure how it would improve the maintainers ability to review PRs necessarily. We all run Directus locally against various different configurations[^1] (that can't be previewed in a single docker compose file), and simply checkout the PR code locally to kick the tires. A live hosted preview link of the prod build would be nice, but won't be able to replace our local environments that are running in dev mode (which allows us to quickly tweak some code and test the differences). 

[^1]: different databases, storage adapters, etc etc)

**Reply by waveywaves** [reply]
> Definitely an interesting concept, but I'm not sure how it would improve the maintainers ability to review PRs necessarily

Before pull the code and testing it locally, you could check the integrations in the preview environment to see if directus works as it is supposed to with the database instances hosted alongside it.

> A live hosted preview link of the prod build would be nice, but won't be able to replace our local environments that are running in dev mode (which allows us to quickly tweak some code and test the differences). 

I can try exploring the possibility of what a preview environment looks like for Directus maintainers. What all tools do you use in your day to day for testing directus (commandline tools or any other tools) ?

**Reply by rijkvanzanten** [reply]
> Before pull the code and testing it locally, you could check the integrations in the preview environment to see if directus works as it is supposed to with the database instances hosted alongside it.

Yeah! I get the idea, but I don't really see how that'll speed our workflow up for Directus specifically. For most internal corporate projects where you have a single stack to test, I can imagine that having a Netlify-esque preview link is incredibly useful, but for our Directus' PR review approach I just don't see the added benefit at this point. 

> I can try exploring the possibility of what a preview environment looks like for Directus maintainers. What all tools do you use in your day to day for testing directus (commandline tools or any other tools) ?

A running prod preview of a PR is not the same as how we'd locally kick the tires. For a PR review, we'd normally wait for the CI to complete the first tests (eg static/unit/e2e tests), pull the branch local, run the affected parts in dev mode, and then start kicking the tires against various configuration options (eg different databases/caches/filestores/etc). An important part there is to change the code of the PR on the fly and see what the output is. 

If we'd hypothetically want to be able to swap out that workflow in favor of a hosted preview, we would have to run more or less all permutations of the integrations and env vars, which is untenable for previews.

I'm also not sure how Uffizzi would handle credentials in this case. We'd either not be able to login to test the app (cause we don't know the creds), or the access is public in which case previews will get spammed and abused in no-time 🤔
